src/crimson/os/seastore/transaction_manager.h

```diff
@@ -335,6 +335,75 @@ class TransactionManager : public ExtentCallbackInterface {
     });
   }
 
+  /**
+   * remap_pin
+   *
+   * Remap original extent to new extent.
+   * Return the pin of new extent.
+   */
+  using remap_pin_iertr = base_iertr;
+  using remap_pin_ret = remap_pin_iertr::future<LBAPinRef>;
+  template <typename T>
+  remap_pin_ret remap_pin(
+    Transaction &t,
+    LBAPinRef &&pin,
+    extent_len_t remap_offset,
+    extent_len_t remap_len) {
+    LOG_PREFIX(TransactionManager::remap_pin);
+    auto remap_laddr = pin->get_key() + remap_offset;
+    auto remap_paddr = pin->get_val().add_offset(remap_offset);
+    // FIXME: paddr can be absolute and pending
+    ceph_assert(pin->get_val().is_absolute());
+    ceph_assert(remap_offset < pin->get_length());
+    ceph_assert(remap_offset + remap_len <= pin->get_length());
+
+    SUBDEBUGT(seastore_tm, "remapping on {} side,"
+      " original laddr: {}, original paddr: {}, original length: {}"
+      " remap laddr: {}, remap paddr: {}, remap length: {}", t,
+      remap_offset == 0 ? "left" : "right",
+      pin->get_key(), pin->get_val(), pin->get_length(),
+      remap_laddr, remap_paddr, remap_len);
+    return cache->get_extent_if_cached(
+      t, pin->get_val(), T::TYPE
+    ).si_then([this, &t, original_paddr = pin->get_key(),
+              remap_laddr, remap_paddr, remap_len](auto ext) {
+      std::optional<ceph::bufferptr> bp;
+      if (ext) {
+        ceph_assert(ext->get_length() > remap_len);
+        bp = ext->get_bptr();
+      } else {
+        bp = std::nullopt;
+      }
+      return alloc_remapped_extent_with_retire<T>(
+        t,
+        remap_laddr,
+        remap_paddr,
+        remap_len,
+        original_paddr,
+        std::move(bp)
+      ).handle_error_interruptible(
+          remap_pin_iertr::pass_further{},
+          crimson::ct_error::assert_all{
+            "TransactionManager::remap_pin hit invalid error"
+          }
+      ).si_then([&t, remap_laddr](auto ext) {
+        if (ext->get_laddr() != remap_laddr) {
+          LOG_PREFIX(TransactionManager::remap_pin);
+          SUBERRORT(
```

Comment:
Do we need this if {}? It seems that there is ceph_assert(ext->get_laddr() == remap_laddr); Is this just for debug?

Reply:
Yes that's unnecessary, I remove that.